ape_solidity/compiler.py

```diff
@@ -738,6 +739,33 @@ def enrich_error(self, err: ContractLogicError) -> ContractLogicError:
                 txn=err.txn,
             )
 
+    def _flatten_source(self, path: Path, base_path=None) -> str:
+        base_path = base_path or path.parent
+        imports = self.get_imports([path])
+        source = ""
+        for import_list in imports.values():
+            if not import_list:
+                continue
+            for import_path in import_list:
+                source += self._flatten_source(base_path / import_path, base_path=base_path)
+        source += path.read_text()
+        return source
+
+    def flatten_contract(self, path: Path, **kwargs) -> Content:
+        """Flatten a contract.
+
+        Args:
+            path: Path to contract file.
+            **kwargs: Keyword arguments
+
+        Returns:
+            Content of flattened contract.
+        """
+        source = self._flatten_source(path)
```

Comment:
took a lookie at this thing: https://github.com/BlockCatIO/solidity-flattener/blob/development/flattener/core.py

is there more we have to do here?

Reply:
we have access to the AST, i made sure to add it in anticipation we would need it here.

if we go with a simpler route, we at least need to make sure there are not multiple SPDX identifiers, i think that will fail polygonscan

Reply:
ah I see, I've mainly used hardhat's flattener in the past which does things like leave multiple SPDX identifiers

from looking at the flattener code you linked, looks like these are the checks being done:

1. Contracts with the same name being defined multiple times in the imports
2. Various forms of "missing source for required dependency"

Compilation should fail if either of these cases happen, so what if we just try compiling the target file first before flattening? Assuming that succeeds, we should be able to move on to flattening (almost) as-is. If it fails, the compilation error message should already be descriptive enough to alert as to the user as to why, and would also prevent us having to implement what is in essence "compiler logic" in the plugin. 

on the issue of SPDX identifiers

We can really only remove some if all of the identifiers across all the files are the same. Otherwise we'd be forced to either fall back to leaving multiple licenses anyway, or choosing one to leave, or removing them all. 

leaving multiple licenses does make it fail on polygonscan and the like, but at least it forces the user to make a decision about their license problem. 

we could fail if more than one license is present, but that's not great UX

I don't think making the plugin choose one license to leave and one to remove is a good idea

removing *all* the SPDX identifiers would at least treat all licenses equally?

what do you think @antazoey ?

Reply:
Ah ok, thanks for diving into that. I agree, anything the compiler validates, we don't need to again.

For the multiple SPDX, I thinking failing early is good. Why do you say that is bad UX? I think if it failed with a nice message  saying there is license discrepancy in the project, that the user should resolve that before trying to validate, that seems good. A project with multiple licenses is indeed invalid.

TBH, I am good with however this works such that Polygonscan now works. So if you are able to flatten a normal-sized contract and use it Polygonscan, we goooood.